kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/yarn/EngineYarnModeSubmitter.scala

```diff
@@ -0,0 +1,477 @@
+        }
+        throw new KyuubiException("Failed to submit application to YARN", e)
+    } finally {
+      if (yarnClient != null) {
+        yarnClient.stop()
+      }
+    }
+  }
+
+  private def createContainerLaunchContext(): ContainerLaunchContext = {
+    info("Setting up container launch context for engine AM")
+    val env = setupLaunchEnv(kyuubiConf)
+    val localResources = prepareLocalResources(stagingDirPath, env)
+
+    val amContainer = Records.newRecord(classOf[ContainerLaunchContext])
+    amContainer.setLocalResources(localResources.asJava)
+    amContainer.setEnvironment(env.asJava)
+
+    val javaOpts = ListBuffer[String]()
+
+    val javaOptions = kyuubiConf.get(ENGINE_DEPLOY_YARN_MODE_JAVA_OPTIONS)
+    if (javaOptions.isDefined) {
+      javaOpts += javaOptions.get
+    }
+
+    val am = Seq(applicationMaster)
+
+    val engineClass = Seq("--class", engineMainClass())
+
+    val kyuubiConfProperties = Seq(
+      "--properties-file",
+      buildPath(Environment.PWD.$$(), LOCALIZED_CONF_DIR, KYUUBI_CONF_FILE))
+
+    val commands =
+      Seq(Environment.JAVA_HOME.$$() + "/bin/java", "-server") ++
+        javaOpts ++ am ++ engineClass ++ kyuubiConfProperties ++
+        Seq(
+          "1>",
+          ApplicationConstants.LOG_DIR_EXPANSION_VAR + "/stdout",
+          "2>",
+          ApplicationConstants.LOG_DIR_EXPANSION_VAR + "/stderr")
+
+    val printableCommands = commands.map(s => if (s == null) "null" else s).toList
+    amContainer.setCommands(printableCommands.asJava)
+    info(s"Commands: ${printableCommands.mkString(" ")}")
+
+    amContainer
+  }
+
+  private def prepareLocalResources(
+      destDir: Path,
+      env: mutable.HashMap[String, String]): mutable.HashMap[String, LocalResource] = {
+    info("Preparing resources for engine AM container")
+    // Upload kyuubi engine and the extra JAR to the remote file system if necessary,
+    // and add them as local resources to the application master.
+    val fs = destDir.getFileSystem(hadoopConf)
+
+    val localResources = mutable.HashMap[String, LocalResource]()
+    FileSystem.mkdirs(fs, destDir, new FsPermission(STAGING_DIR_PERMISSION))
+
+    distributeJars(localResources, env)
+    distributeConf(localResources)
+    localResources
+  }
+
+  private def distributeJars(
+      localResources: mutable.HashMap[String, LocalResource],
+      env: mutable.HashMap[String, String]): Unit = {
+    val jarsArchive = File.createTempFile(LOCALIZED_LIB_DIR, ".zip", Utils.createTempDir().toFile)
+    val jarsStream = new ZipOutputStream(new FileOutputStream(jarsArchive))
+    try {
+      jarsStream.setLevel(0)
+      val jars = kyuubiConf.getOption(KYUUBI_ENGINE_DEPLOY_YARN_MODE_JARS_KEY)
+      assert(jars.isDefined, "No jars specified for engine AM")
+      val putedEntry = new ListBuffer[String]
+      jars.get.split(KYUUBI_ENGINE_DEPLOY_YARN_MODE_ARCHIVE_SEPARATOR).foreach { path =>
+        val jars = Utils.listFilesRecursive(new File(path))
```

Comment:
it does not work well when user provides `HIVE_HADOOP_CLASSPATH`, we should cover the following cases:

- /opt/hadoop/share/hadoop/client/*
- /opt/hadoop/share/hadoop/client/hadoop-client-api-3.3.6.jar:/opt/hadoop/share/hadoop/client/hadoop-client-runtime-3.3.6.jar

Reply:
Oh, I missing this point, that's because my design deals with local paths rather than classpath, and I plan to support external classpath in future PR, where users can customize the archive to be uploaded to the AM's working directory, and then we parse it into classpath.

Reply:
now it's kind of a blocking issue for out-of-box usage, as the hive tgz does not ship hadoop jars in default